design-documents/service-isolation/caching-layer.md

```diff
@@ -0,0 +1,105 @@
+# Caching layer for Magento services
+
+## Overview
+
+Cache is one of the most important parts in a modern web application architecture.
+Magento caching layer should effectively improve:
+ * Performance
+ * Availability
+ * Scalability
+
+Caching should be implemented on several levels:
+ * Caching of http-responses returned by [BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends) (exposed web-server) 
+ * Application data caching (results of DB queries, merged configurations etc.)
+
+### Caching of http-responses returned by BFF (exposed web-server)
+
+[BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends) http-responses contains static assets, **public** and **private** dynamic content.
+Static assets should be cached using CDN. Reverse proxy should cache public content.
+It could be Varnish (acts as reverse proxy) or Fastly (combines reverse proxy and CDN functions).
+Varnish and Fastly are already integrated with Magento. Reverse proxy uses lazy caching approach.
+Beside performance this type of cache has next benefits:
+ * Protection against outages - can optionally serve stale content when there is a problem with origin server
+ * Scalability - number of caching nodes can be increased
+ * Flexibility – Varnish Configuration Language (VCL) builds customized solutions, rules and modules
+ * Shielding - the shield node is the only node to get the content from origin server, other cache nodes will fetch from the shield node
+ 
+ ![Public content caching](img/public-cache.jpg)
+
+A private or visitor-specific content should be stored on client-side (browser). After POST requests the cache 
+in the web browser will be flushed and an AJAX call will be done to fetch an updated copy of the private content.
+Also private content cache on client-side should expire according to TTL.
+
+ ![Private content caching](img/private-cache.jpg)
+ 
+ In general, the same caching approach works in current monolith Magento architecture.
+ For service oriented architecture next issues should be solved:
+  * The http-response header should include a set of tags for effective cache purging on reverse-proxy.
```

Comment:
What would be an example of request performed by multiple services?

Reply:
So it's a case when GET request after BFF goes to service1 and service2 performs a request to service2 for some data.
For example this one goes to Checkout service and than request data from Payment service:

` <route url="/V1/guest-carts/:cartId/payment-information" method="GET">
        <service class="Magento\Checkout\Api\GuestPaymentInformationManagementInterface" method="getPaymentInformation"/>
        <resources>
            <resource ref="anonymous" />
        </resources>
    </route>`